src/rulesets/Base/array.jl

```diff
@@ -43,32 +51,81 @@ function rrule(::typeof(Base.vect), X::Vararg{Any,N}) where {N}
     return Base.vect(X...), vect_pullback
 end
 
+"""
+    _make_real_zeros(xdots, xs)
+
+Forward rules for `vect` or `cat` may receive a mixture of data and `ZeroTangent`s.
+To avoid `vect(1, ZeroTangent(), 3)` or `hcat([1,2], ZeroTangent())`, this materialises
+each zero `xdot` to be `zero(x)`.
+"""
+_make_real_zeros(xdots, xs) = map(_real_zero, xdots, xs)
+_real_zero(xdot, x) = xdot
+_real_zero(xdot::AbstractZero, x) = zero(x)
+
+# Fast paths. Should it also collapse all-Zero cases?
+_make_real_zeros(xdots::NTuple{<:Any, <:Number}, xs) = xdots
+_make_real_zeros(xdots::AbstractArray{<:Number}, xs) = xdots
+_make_real_zeros(xdots::AbstractArray{<:AbstractArray}, xs) = xdots
```

Comment:
Should we use
`eltype` and `HasEltype`  here? so that we are abstracted over collections?
Possibly not required right now and we can leave it?

Reply:
My guess is that this is already into premature optimisation territory. A tuple of numbers for `vect`, and a tuple of arrays for `vcat`, are the cases that I actually managed to trigger. Maybe there should be a `xdots::NTuple{<:Any, <:AbstractArray}` method. Although `map` on tuples is mostly free.